docs/adr/009_lns_affinity.md

```diff
@@ -0,0 +1,244 @@
+Edge Hub on LNS1, will detect this and assume it's a transient network issue, therefore will try
+proactively to reconnect to IoT Hub. IoT Hub will now drop the connection 2 to re-establish the
+original connection 1.  
+
+This connection "ping-pong" will continue happening, negatively impacting the scalability due to the
+high costs of setting up/disposing the connections. From our load tests we observed that in this
+scenario we were not even able to connect more than 120 devices to two LNSs, while in a single LNS
+topology we could scale up to 900 devices without issues.
+
+## Out of scope
+
+- Deduplication strategies Mark and None: these strategies rely on multiple LNSs sending message.
+Potentially we could consider other work arounds for the IoT Hub limitation of a single connection
+per device but we find it acceptable for the Mark and None strategies to not be as scalable as the
+Drop strategy and will only document this limitation for potential users to be aware of.
+
+- LNS performs operations on behalf of a device/sensor and a concentrator/station. However since a
+concentrator can be connected to at most one LNS, there is no ping-pong happening with operations on
+stations.
+
+## Limitations
+
+- We should avoid introducing additional calls to the Function as this would also hurt scalability.
+  
+## In-scope
+
+- The problem can be manifested whenever we do operations against Iot Hub on behalf of edge devices.
+  These can be:
+  - Twin reads
+  - Twin writes (updates/deletes)
+  - D2C messages
+  - C2D messages
+- Roaming edge devices (that potentially lose connectivity to an LNS) are in scope.
+
+### Problematic connections opened on behalf of edge devices
+
+- Data message flows
+  - Join: 
+  - Data: MessageDispatcher.DispatchLoraDataMessage gets the twin (if LoRaDevice not in cache)
+    - Direct method for Class C should return an error if it could not send
+      message downlink so that the user can retry.
+
+Version, LNS discovery and CUPS update endpoints are not affected.
+
+### Background tasks
+
+- LoRaDeviceCache refresh in the background: every 10 minutes we are checking for devices that need
+  refreshing (those that were not refreshed in the last 2 days). Refreshing a device gets the twin
+  of the device ❔
+  - if we are the losing LNS we adjust LastSean property but not refresh the entry
+  - in order to not have stale entries we could refresh the twin once we become the winning LNS from
+    the data (also required for join?) flow  
+
+Alternative:
+SearchByDevAddr: the function returns info about the winning gateway
+Requires doing the Mic computation on the Function
+
+## Possible solutions
+
+### Delayed processing of messages from losing LNSs
+
+The main idea here is to delay the processing of future messages for all gateways *besides* the
+winning one. This should give enough time to the chosen LNS to process the message and keep the
```

Comment:
I would formulate this a bit differently. We are giving the current connection owner an edge, to win the processing race. That gateway, will have 0 impact on performance.

Reply:
Reworded.